docs/dev/guidelines/criteria-builder.rst

```diff
@@ -0,0 +1,294 @@
+
+    .. code-block:: java
+
+        public interface Specification<T> {
+          Predicate toPredicate(Root<T> root, CriteriaQuery query, CriteriaBuilder cb);
+        }
+
+    Now we can create Specifications. We can achieve this in two ways:
+
+    - Anonymous new Specification<User>():
+
+        .. code-block:: java
+
+            private Specification<User> getAllUsersMatchingEmptyCourses() {
+                return new Specification<User>() {
+                     @Override
+                     public Predicate toPredicate(Root<User> root, CriteriaQuery<?> query, CriteriaBuilder criteriaBuilder) {
+                        return criteriaBuilder.isEmpty(root.get(User_.GROUPS));
+                    }
+                };
+            }
+
+    - Lambda expression (preferred version):
+
+        .. code-block:: java
+
+            private Specification<User> getAllUsersMatchingEmptyCourses() {
+            	return (root, query, criteriaBuilder) -> criteriaBuilder.isEmpty(root.get(User_.GROUPS));
+            }
+
+
+3. Operations
+=============
+
+- **AND:** We can perform the `and` operation on an arbitrary number of predicates via the criteriaBuilder object, which results in a new Predicate.
+
+    .. code-block:: java
+
+        return (root, query, criteriaBuilder) -> {
+        	Predicate one = criteriaBuilder.equal(x, z);
+        	Predicate two = criteriaBuilder.notEqual(a, b);
+
+        	return criteriaBuilder.and(one, two, ...);
+        };
+
+- **OR:** We can perform the `or` operation on an arbitrary number of predicates via the criteriaBuilder object, which results in a new Predicate.
+
+    .. code-block:: java
+
+        return (root, query, criteriaBuilder) -> {
+            Predicate one = criteriaBuilder.equal(x, z);
+            Predicate two = criteriaBuilder.notEqual(a, b);
+
+            return criteriaBuilder.or(one, two, ...);
+        };
+
+- **EQUAL / NOT EQUAL:**
+
+    .. code-block:: java
+
+        return (root, query, criteriaBuilder) -> {
+         	Predicate one = criteriaBuilder.equal(root.get(User_.IS_INTERNAL), true);
+        	Predicate two = criteriaBuilder.notEqual(root.get(User_.ACTIVATED), true);
+
+        	return criteriaBuilder.and(one, two, ...);
+        };
+
+- **NOT:**
+
+    .. code-block:: java
+
+        return (root, query, criteriaBuilder) -> {
+            ...
+            Predicate predicate = criteriaBuilder.exists(subQuery).not();
+
+            return criteriaBuilder.equals(predicate);
+        };
+
+- **IN:** To check if the collection contains a value.
+
+    .. code-block:: java
+
+        return (root, query, criteriaBuilder) -> {
+         	Predicate in = criteriaBuilder.in(root.get(User_.ID)).value(ids);
+            return in;
+        };
+
+
+4. Joins
+========
+
+Many different joins are available (e.g. Join, ListJoin, SetJoin, CollectionJoin, ...) - please choose the right one.
```

Comment:
```suggestion
Different joins are available (e.g. Join, ListJoin, SetJoin, CollectionJoin, ...) - please choose the right one.
```
Nitpick: one of {many, different} is technically redundant. Keeping just ‘different’ highlights that the joins implement different operations.